hal/src/sercom/uart/pads_thumbv7em.rs

```diff
@@ -1,91 +1,156 @@
+};
+use core::marker::PhantomData;
 
 //=============================================================================
-// Rxpo
+// RxpoTxpo
 //=============================================================================
 
-/// Control the `RXPO` field as a function of the [`PadNum`] type
-pub trait Rxpo: Sealed {
-    /// Corresponding variant from the PAC `enum`
-    const RXPO: RXPO_A;
-}
-
-impl Rxpo for Pad0 {
-    const RXPO: RXPO_A = RXPO_A::PAD0;
-}
-impl Rxpo for Pad1 {
-    const RXPO: RXPO_A = RXPO_A::PAD1;
-}
-impl Rxpo for Pad2 {
-    const RXPO: RXPO_A = RXPO_A::PAD2;
-}
-impl Rxpo for Pad3 {
-    const RXPO: RXPO_A = RXPO_A::PAD3;
-}
-
-impl Rxpo for NoneT {
-    /// This value is arbitrary and meaningless for [`NoneT`]
-    const RXPO: RXPO_A = RXPO_A::PAD0;
+/// Configure the `RXPO` and `TXPO` fields based on a set of [`Pads`]
+///
+/// According to the datasheet, the `RXPO` and `TXPO` values specify which
+/// SERCOM pads are used for various functions. Moreover, depending on which
+/// pads are actually in use, only certain combinations of these values make
+/// sense and are valid.
+///
+/// This trait is implemented for valid, four-tuple combinations of
+/// [`OptionalPadNum`]s. Those implementations are then lifted to the
+/// corresponding [`Pads`] types.
+///
+/// To satisfy this trait, the combination of [`OptionalPadNum`]s must specify
+/// [`PadNum`] for at least one of `RX` and `TX`. Furthermore, no
+/// two [`PadNum`]s can conflict.
+pub trait RxpoTxpo {
+    /// `RXPO` field value
+    const RXPO: u8;
+
+    /// `RXPO` field value
+    const TXPO: u8;
 }
 
-/// Lift the implementations of [`Rxpo`] from [`OptionalPadNum`]s to the
-/// corresponding [`Pads`] types.
-impl<S, I, RX, TX, RTS, CTS> Rxpo for Pads<S, I, RX, TX, RTS, CTS>
+/// Lift the implementations of [`RxpoTxpo`] from four-tuples of
+/// [`OptionalPadNum`]s to the corresponding [`Pads`] types.
+impl<S, I, RX, TX, RTS, CTS> RxpoTxpo for Pads<S, I, RX, TX, RTS, CTS>
 where
     S: Sercom,
     I: IoSet,
     RX: OptionalPad,
     TX: OptionalPad,
     RTS: OptionalPad,
     CTS: OptionalPad,
-    RX::PadNum: Rxpo,
+    (RX::PadNum, TX::PadNum, RTS::PadNum, CTS::PadNum): RxpoTxpo,
 {
-    const RXPO: RXPO_A = RX::PadNum::RXPO;
+    const RXPO: u8 = <(RX::PadNum, TX::PadNum, RTS::PadNum, CTS::PadNum)>::RXPO;
+    const TXPO: u8 = <(RX::PadNum, TX::PadNum, RTS::PadNum, CTS::PadNum)>::TXPO;
 }
 
 //=============================================================================
-// Txpo
+// Implement RxpoTxpo
 //=============================================================================
 
-/// Control the `TXPO` field as a function of the [`PadNum`] type
-pub trait Txpo: Sealed {
-    /// Corresponding variant from the PAC `enum`
-    const TXPO: TXPO_A;
-}
-
-impl Txpo for Pad0 {
-    const TXPO: TXPO_A = TXPO_A::TXPO_2;
-}
+/// Filter [`PadNum`] permutations and implement [`RxpoTxpo`]
+    #[rustfmt::skip]
+    macro_rules! impl_rxpotxpo {
```

Comment:
Shouldn't this be indented less?

Reply:
Ah yeah indeed. I'll fix that tomorrow, then I think we can merge